### PR TITLE
Rework the average active addresses to use Clickhouse instead of TimescaleDB

### PR DESCRIPTION
**Pull request by IvanIvanoff**
Add support for Bitcoin average active addresses in the all projects
queries

#### Summary

[//]: # (#### Related PRs)

[//]: # (#### Additional deploy notes)

[//]: # (#### Screenshots)

### CODE REVIEW

File: lib/sanbase/clickhouse/bitcoin/bitcoin.ex [hunk 1]
```diff
@@ -111,6 +115,31 @@ defmodule Sanbase.Clickhouse.Bitcoin do
     {query, args}
   end
 
+  defp get_aggregated_metric_in_interval(metric, aggregation, from, to)
+       when metric in @metrics and aggregation in @aggregations do
+    {query, args} = get_aggregated_metric_in_interval_query(metric, aggregation, from, to)
+
+    ClickhouseRepo.query_transform(query, args, fn [value] -> value end)
+    |> case do
+      {:ok, [result]} -> {:ok, result}
```
Comment (tspenov) [inline lib/sanbase/clickhouse/bitcoin/bitcoin.ex]: it it list here `[result]` ?

File: lib/sanbase/clickhouse/daily_active_addresses.ex [hunk 2]
```diff
@@ -2,32 +2,72 @@ defmodule Sanbase.Clickhouse.DailyActiveAddresses do
...
+        c, {btc, eth, erc20} when c in @ethereum -> {btc, [c | eth], erc20}
+        c, {btc, eth, erc20} when is_binary(c) -> {btc, eth, [c | erc20]}
+        _, acc -> acc
+      end)
+
+    with {:ok, btc_average_addresses} <- do_btc_average_active_addresses(btc, from, to),
+         {:ok, eth_average_addresses} <- do_eth_average_active_addresses(eth, from, to),
+         {:ok, erc20_average_addresses} <- do_erc20_average_active_addresses(erc20, from, to) do
+      {:ok, btc_average_addresses ++ eth_average_addresses ++ erc20_average_addresses}
+    end
```
Comment (tspenov) [inline lib/sanbase/clickhouse/daily_active_addresses.ex]: Very nice implementation! 🎉

File: test/sanbase_web/graphql/clickhouse/average_daily_active_addresses_api_test.exs [hunk 3]
```diff
@@ -0,0 +1,112 @@
...
+    project = insert(:project, %{main_contract_address: "0x123123"})
+    insert(:project, %{coinmarketcap_id: "ethereum", ticker: "ETH", name: "Ethereum"})
+    insert(:project, %{coinmarketcap_id: "bitcoin", ticker: "BTC", name: "Bitcoin"})
+    datetime1 = DateTime.from_naive!(~N[2017-05-13 00:00:00], "Etc/UTC")
+    datetime2 = DateTime.from_naive!(~N[2017-05-23 00:00:00], "Etc/UTC")
+
+    [
+      project: project,
+      datetime1: datetime1,
+      datetime2: datetime2
```
Comment (kalkov) [inline test/sanbase_web/graphql/clickhouse/average_daily_active_addresses_api_test.exs]: "to" is better

File: test/sanbase_web/graphql/clickhouse/average_daily_active_addresses_api_test.exs [hunk 4]
```diff
@@ -0,0 +1,143 @@
...
+        |> post("/graphql", query_skeleton(query, "projectBySlug"))
+
+      active_addresses =
+        json_response(result, 200)["data"]["projectBySlug"]["averageDailyActiveAddresses"]
+
+      assert active_addresses == 0
+    end
+  end
+
+  test "average daily active addresses works for ethereum and bitcoin", context do
```
Comment (kalkov) [inline test/sanbase_web/graphql/clickhouse/average_daily_active_addresses_api_test.exs]: test "returns average DAA for ERC20 tokens, ethereum and bitcoin requested together"

File: test/sanbase_web/graphql/clickhouse/average_daily_active_addresses_api_test.exs [hunk 5]
```diff
@@ -0,0 +1,143 @@
...
+    datetime2 = DateTime.from_naive!(~N[2017-05-23 00:00:00], "Etc/UTC")
+
+    [
+      project: project,
+      datetime1: datetime1,
+      datetime2: datetime2
+    ]
+  end
+
+  test "average daily active addreses for projectBySlug", context do
```
Comment (kalkov) [inline test/sanbase_web/graphql/clickhouse/average_daily_active_addresses_api_test.exs]: Can you add also a test for projectBySlug + bitcoin?

File: lib/sanbase/clickhouse/daily_active_addresses.ex [hunk 6]
```diff
@@ -2,32 +2,90 @@ defmodule Sanbase.Clickhouse.DailyActiveAddresses do
...
+  # In case of error return `{:ok, []}` because the other 2 queries could succeed
+  # and return meaningful results
+  defp do_btc_average_active_addresses([], _, _), do: {:ok, []}
+
+  defp do_btc_average_active_addresses([_ | _], from, to) do
+    with {:ok, result} <- Bitcoin.average_active_addresses(from, to) do
+      {:ok, [{"BTC", result}]}
+    else
+      {:error, error} ->
+        Logger.warn("Cannot fetch average active addresses for Bitcoin")
```
Comment (kalkov) [inline lib/sanbase/clickhouse/daily_active_addresses.ex]: You can extract this logging into a private function and pass the name of the crypto. 
Maybe the text should be "Cannot fetch average daily active addresses for ..."?

File: lib/sanbase/clickhouse/daily_active_addresses.ex [hunk 7]
```diff
@@ -2,32 +2,90 @@ defmodule Sanbase.Clickhouse.DailyActiveAddresses do
...
+      when is_binary(eth) and eth in @ethereum do
     Eth.average_active_addresses(from, to, interval)
   end
 
-  def average_active_addresses(contract, from, to, interval) do
+  def average_active_addresses(contract, from, to, interval) when is_binary(contract) do
     Erc20.average_active_addresses(contract, from, to, interval)
   end
+
+  def average_active_addresses(contracts, from, to) when is_list(contracts) do
```
Comment (kalkov) [inline lib/sanbase/clickhouse/daily_active_addresses.ex]: Art :framed_picture:

File: lib/sanbase/clickhouse/bitcoin/bitcoin.ex (hunk 1 above)
Comment (IvanIvanoff) [reply lib/sanbase/clickhouse/bitcoin/bitcoin.ex]: Yes.

**Pull request merged**